### DISCUSSION TITLE
TimeSeries pagination

### DISCUSSION THREAD

**Discussion by rripken** [USACE/cwms-data-api#78] [General]
Is pagination (page, nextPage, total, pageSize) staying inside TimeSeries?  Or should it be moved above or next-to the TimeSeries object in the results?

**Comment by MikeNeilson** [comment]
We had talked about that but haven't made a decission yet.

My vote would be in favor of separate. the pagination object hold an otherwise complete DTO.  However, I'm not sure quite how well this works for things like Rating tables. Arguably it's not super difficult to recombine, and that COULD be handled as part of the cwms-radar-client that's being created. But it is a concern. It's also possible we'll just use a different mechanism for the ratings and don't need to paginate them.

**Reply by adamkorynta** [reply]
I agree with separating them for time series. I haven't dug into the ratings API yet so I don't know how that will work.

**Reply by MikeNeilson** [reply]
presumably it would work the same way. However, I think it is conceptually easier for must to understand how to stitch together a time series vs paired data.

**Reply by MikeNeilson** [reply]
I will investigate how to separate out the DTO of interest from the pagination. But my gut reaction is something like the follow 

```java
...
public class PaginatedOutput<T> {
    private Page page;
    private T data;
    ... etc
}

...
var output = new PaginatedOutput<TimeSeries>( The Timeseries, <existing page info>);
...
```

the resulting output (json as example)
```json
{  "page": {
       "current-page": "<opaque handle>",
       "next-page": "<opaque handle>",
      <previous page? if available?>
   }
   "data": { could be a list of objects, could be object}
}
```

Though we'll need to experiment with how the templates affect the swagger schema. Seems to have worked for the existing system.

**Comment by rripken** [comment]
There are several classes that inherit from CwmsDTOPaginated (e.g. Pools, Clobs, Blobs, Catalog, TimeSeries ).    All of those classes, except for TimeSeries, contains an inner list of some DTO type.   Clobs, for example, has a List<Clob> clobs field.   The Clob DTO doesn't know or care about pagination.
I think, to fit the current pattern and not be a sore thumb, TimeSeries would just not extend CwmsDTOPaginated and instead there would need to be a TimeSeriesList class or PagedTimeSeries  class that did extend CwmsDTOPaginated and had a List<TimeSeries>.   It would just make the class consistent with the others and make it so the base TimeSeries DTO was oblivious to paging.

**Reply by MikeNeilson** [reply]
That makes sense. and would certainly be easier than rewriting the whole thing.
